File: agent-lee-coding-mode/runtime/discovery_loader.py
import json
import os
from typing import Optional, List, Dict, Any

def _strip_bom(raw: str) -> str:
    return str(raw or "").lstrip("\ufeff")

def _repo_root() -> str:
    here = os.path.dirname(__file__)
    return os.path.abspath(os.path.join(here, '..'))

def _workspace_root() -> str:
    return os.path.abspath(os.path.join(_repo_root(), '..'))

def _index_candidates() -> List[str]:
    root = _repo_root()
    candidates = [
        os.path.join(root, 'agent-lee-coding-mode', 'source-index', 'all-leeway-files.index.json.new'),
        os.path.join(root, 'agent-lee-coding-mode', 'source-index', 'all-leeway-files.index.json'),
    ]
    return candidates

def _standards_root_map_candidates() -> List[str]:
    workspace = _workspace_root()
    return [
        os.path.join(workspace, 'LeeWay-Standards', 'standards-root-map.json'),
        os.path.join(workspace, 'Archive', 'docs', 'deprecated-standards-roots', 'Leeway-Standards', 'standards-root-map.json'),
    ]

def _normalize_standards_root_map(map_obj: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    if not isinstance(map_obj, dict):
        return None
    archived_roots = map_obj.get('archivedRoots') if isinstance(map_obj.get('archivedRoots'), list) else []
    return {
        'canonicalRoot': map_obj.get('canonicalRoot') if map_obj.get('canonicalRoot') else 'LeeWay-Standards',
        'archivedRoots': [r for r in archived_roots if r],
        'mergeTimestamp': map_obj.get('mergeTimestamp'),
        'filesMerged': map_obj.get('filesMerged', 0) if isinstance(map_obj.get('filesMerged', 0), int) else 0,
        'conflictsResolved': map_obj.get('conflictsResolved', 0) if isinstance(map_obj.get('conflictsResolved', 0), int) else 0,
        'conflictsRemaining': map_obj.get('conflictsRemaining', 0) if isinstance(map_obj.get('conflictsRemaining', 0), int) else 0,
        'writePolicy': map_obj.get('writePolicy') if map_obj.get('writePolicy') else 'All new standards writes must go to LeeWay-Standards.',
        'discoveryPolicy': map_obj.get('discoveryPolicy') if map_obj.get('discoveryPolicy') else 'Only canonical root is active. Archived roots are excluded from runtime discovery.',
        'receiptRequired': map_obj.get('receiptRequired', True) is not False,
    }

def _inspect_standards_root_state(root_map: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    workspace = _workspace_root()
    canonical_root = os.path.join(workspace, 'LeeWay-Standards')
    deprecated_root = os.path.join(workspace, 'Leeway-Standards')
    archived_roots = []
    if root_map and isinstance(root_map.get('archivedRoots'), list) and root_map['archivedRoots']:
        archived_roots = list(root_map['archivedRoots'])
    else:
        archived_roots = [os.path.join('Archive', 'docs', 'deprecated-standards-roots', 'Leeway-Standards')]

    warnings: List[str] = []
    canonical_root_exists = os.path.exists(canonical_root)
    canonical_real = os.path.realpath(canonical_root) if canonical_root_exists else None
    deprecated_real = os.path.realpath(deprecated_root) if os.path.exists(deprecated_root) else None
    deprecated_root_exists = bool(deprecated_real and canonical_real and deprecated_real.lower() != canonical_real.lower())
    archived_root_exists = os.path.exists(os.path.join(workspace, archived_roots[0])) if archived_roots else False

    if not canonical_root_exists:
        warnings.append(f'canonical standards root missing: {canonical_root}')
    if deprecated_root_exists:
        warnings.append(f'deprecated standards root still active outside archive: {deprecated_root}')
    if archived_roots and not archived_root_exists:
        warnings.append(f'archived standards root missing: {archived_roots[0]}')
    if root_map and root_map.get('canonicalRoot') and root_map.get('canonicalRoot') != 'LeeWay-Standards':
        warnings.append(f'standards-root-map canonicalRoot mismatch: {root_map.get("canonicalRoot")}')

    return {
        'canonicalRoot': 'LeeWay-Standards',
        'archivedRoots': archived_roots,
        'canonicalRootExists': canonical_root_exists,
        'deprecatedRootExistsOutsideArchive': deprecated_root_exists,
        'archivedRootExists': archived_root_exists,
        'warnings': warnings,
    }


def load_discovery_index() -> Optional[Dict[str, Any]]:
    for p in _index_candidates():
        if os.path.exists(p):
            try:
                with open(p, 'r', encoding='utf-8') as f:
                    raw = _strip_bom(f.read())
                try:
                    parsed = json.loads(raw)
                    if isinstance(parsed, list):
                        parsed_index = {'files': parsed}
                    else:
                        parsed_index = parsed
                    return _attach_standards_root_metadata(parsed_index)
                except json.JSONDecodeError:
                    # attempt to find last JSON object
                    last_open = raw.rfind('{')
                    last_close = raw.rfind('}')
                    if last_open != -1 and last_close > last_open:
                        candidate = raw[last_open:last_close+1]
                        try:
                            return _attach_standards_root_metadata(json.loads(candidate))
                        except Exception:
                            return None
            except Exception:
                continue
    return None

def load_standards_root_map() -> Optional[Dict[str, Any]]:
    for p in _standards_root_map_candidates():
        if os.path.exists(p):
            try:
                with open(p, 'r', encoding='utf-8') as f:
                    parsed = json.loads(_strip_bom(f.read()))
                return _normalize_standards_root_map(parsed)
            except Exception:
                continue
    return None

def _attach_standards_root_metadata(index: Dict[str, Any]) -> Dict[str, Any]:
    root_map = load_standards_root_map()
    root_state = _inspect_standards_root_state(root_map)
    index['standardsRootMap'] = root_map
    index['standardsRootState'] = root_state
    index['activeStandardsRoot'] = root_state['canonicalRoot']
    index['archivedStandardsRoots'] = list(root_state['archivedRoots'])
    index['standardsRootWarnings'] = list(root_state['warnings'])
    return index


def find_files_by_name(name: str) -> List[str]:
    index = load_discovery_index()
    if not index:
        return []
    files = index.get('files') or []
    matches = []
    for f in files:
        fn = f.get('Name') if isinstance(f, dict) else None
        full = f.get('FullName') if isinstance(f, dict) else None
        if fn == name or (full and name in full):
            matches.append(full)
    return matches


def find_model_role(index: Dict[str, Any], role_name: str) -> Optional[str]:
    if not index:
        return None
    if index.get('modelRoles') and index['modelRoles'].get(role_name):
        return index['modelRoles'][role_name]
    # Check runtime fabric model-authority
    root = os.path.abspath(os.path.join(_repo_root(), '..'))
    mab = os.path.join(root, 'Leeway Runtime Fabric', 'capability-registry', 'registry', 'model-authority.json')
    if os.path.exists(mab):
        try:
            with open(mab, 'r', encoding='utf-8') as f:
                parsed = json.loads(_strip_bom(f.read()))
            if parsed.get('modelRoles') and parsed['modelRoles'].get(role_name):
                return parsed['modelRoles'][role_name]
        except Exception:
            pass
    # Look for updated-model-hive-registry.json in index files
    files = index.get('files') or []
    for f in files:
        full = f.get('FullName') if isinstance(f, dict) else None
        if full and full.endswith('updated-model-hive-registry.json') and os.path.exists(full):
            try:
                with open(full, 'r', encoding='utf-8') as fh:
                    parsed = json.loads(_strip_bom(fh.read()))
                if parsed.get('modelRoles') and parsed['modelRoles'].get(role_name):
                    return parsed['modelRoles'][role_name]
            except Exception:
                continue
    return None

File: agent-lee-coding-mode/runtime/discovery-loader.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const AGENT_LEE_ROOT = path.resolve(__dirname, "..");
const WORKSPACE_ROOT = path.resolve(AGENT_LEE_ROOT, "..");
const ACTIVE_STANDARDS_ROOT = path.join(WORKSPACE_ROOT, "LeeWay-Standards");
const DEPRECATED_STANDARDS_ROOT = path.join(WORKSPACE_ROOT, "Leeway-Standards");
const ARCHIVED_STANDARDS_ROOT = path.join(WORKSPACE_ROOT, "Archive", "docs", "deprecated-standards-roots", "Leeway-Standards");
const DISCOVERY_INDEX_CANDIDATES = [
  path.join(AGENT_LEE_ROOT, "source-index", "all-leeway-files.index.json.new"),
  path.join(AGENT_LEE_ROOT, "source-index", "all-leeway-files.index.json"),
  path.join(AGENT_LEE_ROOT, "source-index", "all-leeway-files.index.json.backup")
];
const GOVERNANCE_MANIFEST_CANDIDATES = [
  path.join(ACTIVE_STANDARDS_ROOT, "runtime-governance-manifest.json")
];
const STANDARDS_ROOT_MAP_CANDIDATES = [
  path.join(ACTIVE_STANDARDS_ROOT, "standards-root-map.json"),
  path.join(ARCHIVED_STANDARDS_ROOT, "standards-root-map.json")
];

function stripBom(raw) {
  return String(raw || "").replace(/^\uFEFF/, "");
}

function tolerantJsonParse(raw) {
  const clean = stripBom(raw);
  try {
    return JSON.parse(clean);
  } catch (err) {
    // Try to recover by locating the last JSON object in the stream
    const lastOpen = clean.lastIndexOf("{");
    const lastClose = clean.lastIndexOf("}");
    if (lastOpen !== -1 && lastClose !== -1 && lastClose > lastOpen) {
      const candidate = clean.substring(lastOpen, lastClose + 1);
      try {
        return JSON.parse(candidate);
      } catch (err2) {
        // Try first-to-last braces
        const firstOpen = clean.indexOf("{");
        if (firstOpen !== -1 && lastClose > firstOpen) {
          try {
            return JSON.parse(clean.substring(firstOpen, lastClose + 1));
          } catch (err3) {
            return null;
          }
        }
        return null;
      }
    }
    return null;
  }
}

function loadFirstJsonCandidate(candidates) {
  for (const candidate of candidates) {
    try {
      if (!fs.existsSync(candidate)) continue;
      const raw = fs.readFileSync(candidate, "utf8");
      const parsed = tolerantJsonParse(raw);
      if (!parsed) continue;
      return { path: candidate, parsed };
    } catch (err) {
      // ignore and continue
    }
  }
  return null;
}

function normalizeStandardsRootMap(map) {
  if (!map || typeof map !== "object") return null;
  const archivedRoots = Array.isArray(map.archivedRoots) ? map.archivedRoots.filter(Boolean) : [];
  return {
    canonicalRoot: typeof map.canonicalRoot === "string" && map.canonicalRoot ? map.canonicalRoot : "LeeWay-Standards",
    archivedRoots,
    mergeTimestamp: typeof map.mergeTimestamp === "string" ? map.mergeTimestamp : null,
    filesMerged: Number.isFinite(map.filesMerged) ? map.filesMerged : 0,
    conflictsResolved: Number.isFinite(map.conflictsResolved) ? map.conflictsResolved : 0,
    conflictsRemaining: Number.isFinite(map.conflictsRemaining) ? map.conflictsRemaining : 0,
    writePolicy: typeof map.writePolicy === "string" ? map.writePolicy : "All new standards writes must go to LeeWay-Standards.",
    discoveryPolicy: typeof map.discoveryPolicy === "string" ? map.discoveryPolicy : "Only canonical root is active. Archived roots are excluded from runtime discovery.",
    receiptRequired: map.receiptRequired !== false
  };
}

function inspectStandardsRootState(rootMap) {
  const warnings = [];
  const archivedRoots = rootMap && Array.isArray(rootMap.archivedRoots) ? rootMap.archivedRoots.slice() : [path.join("Archive", "docs", "deprecated-standards-roots", "Leeway-Standards")];
  const canonicalRootExists = fs.existsSync(ACTIVE_STANDARDS_ROOT);
  const canonicalReal = canonicalRootExists ? fs.realpathSync(ACTIVE_STANDARDS_ROOT) : null;
  const deprecatedReal = fs.existsSync(DEPRECATED_STANDARDS_ROOT) ? fs.realpathSync(DEPRECATED_STANDARDS_ROOT) : null;
  const deprecatedRootExistsOutsideArchive = Boolean(deprecatedReal && canonicalReal && deprecatedReal.toLowerCase() !== canonicalReal.toLowerCase());
  const archivedRootExists = fs.existsSync(path.join(WORKSPACE_ROOT, archivedRoots[0] || ""));

  if (!canonicalRootExists) {
    warnings.push(`canonical standards root missing: ${ACTIVE_STANDARDS_ROOT}`);
  }
  if (deprecatedRootExistsOutsideArchive) {
    warnings.push(`deprecated standards root still active outside archive: ${DEPRECATED_STANDARDS_ROOT}`);
  }
  if (!archivedRootExists && archivedRoots.length > 0) {
    warnings.push(`archived standards root missing: ${archivedRoots[0]}`);
  }
  if (rootMap && rootMap.canonicalRoot && rootMap.canonicalRoot !== "LeeWay-Standards") {
    warnings.push(`standards-root-map canonicalRoot mismatch: ${rootMap.canonicalRoot}`);
  }

  return {
    canonicalRoot: "LeeWay-Standards",
    archivedRoots,
    canonicalRootExists,
    deprecatedRootExistsOutsideArchive,
    archivedRootExists,
    warnings
  };
}

export async function loadAllLeewayIndex() {
  let indexObj = null;
  let indexPath = null;

  for (const candidate of DISCOVERY_INDEX_CANDIDATES) {
    if (!fs.existsSync(candidate)) continue;
    const raw = await fs.promises.readFile(candidate, "utf8");
    const parsed = tolerantJsonParse(raw);
    if (!parsed) continue;
    indexObj = Array.isArray(parsed) ? { files: parsed } : parsed;
    indexPath = candidate;
    break;
  }

  if (!indexObj) return null;
  indexObj.discoveryIndexPath = indexPath;
  indexObj.discoveryIndexCandidates = DISCOVERY_INDEX_CANDIDATES.slice();
  const standardsRootMapEntry = loadFirstJsonCandidate(STANDARDS_ROOT_MAP_CANDIDATES);
  const standardsRootMap = normalizeStandardsRootMap(standardsRootMapEntry && standardsRootMapEntry.parsed);
  const standardsRootState = inspectStandardsRootState(standardsRootMap);
  indexObj.standardsRootMap = standardsRootMap;
  indexObj.standardsRootMapPath = standardsRootMapEntry ? standardsRootMapEntry.path : null;
  indexObj.standardsRootState = standardsRootState;
  indexObj.activeStandardsRoot = standardsRootState.canonicalRoot;
  indexObj.archivedStandardsRoots = standardsRootState.archivedRoots.slice();
  indexObj.standardsRootWarnings = standardsRootState.warnings.slice();

  // Try to load governance manifest from the canonical standards root first.
  for (const govPath of GOVERNANCE_MANIFEST_CANDIDATES) {
    try {
      if (!fs.existsSync(govPath)) continue;
      const govRaw = await fs.promises.readFile(govPath, "utf8");
      const govParsed = tolerantJsonParse(govRaw);
      if (govParsed) {
        indexObj.governanceManifest = govParsed;
        indexObj.governanceManifestPath = govPath;
        break;
      }
    } catch (err) {
      // ignore
    }
  }

  return indexObj;
}

export function findModelRole(index, roleName) {
  if (!index) return null;
  // First, check index-provided modelRoles
  if (index.modelRoles && index.modelRoles[roleName]) return index.modelRoles[roleName];

  // Next, check canonical model authority manifest under Leeway Runtime Fabric (discovery-first)
  try {
    const modelAuthorityPath = path.join(AGENT_LEE_ROOT, '..', 'Leeway Runtime Fabric', 'capability-registry', 'registry', 'model-authority.json');
    if (fs.existsSync(modelAuthorityPath)) {
      const raw = stripBom(fs.readFileSync(modelAuthorityPath, 'utf8'));
      const mab = tolerantJsonParse(raw);
      if (mab && mab.modelRoles && mab.modelRoles[roleName]) return mab.modelRoles[roleName];
    }
  } catch (err) {
    // ignore and continue
  }
  // Try to find an external updated-model-hive-registry reference inside index.files
  if (Array.isArray(index.files)) {
    const hive = index.files.find((f) => {
      const fullName = typeof f === "string"
        ? f
        : f && typeof f.FullName === "string"
          ? f.FullName
          : f && typeof f.fullName === "string"
            ? f.fullName
            : "";
      return fullName.endsWith("updated-model-hive-registry.json");
    });
    if (hive) {
      try {
        const hivePathRaw = typeof hive === "string"
          ? hive
          : hive && typeof hive.FullName === "string"
            ? hive.FullName
            : hive && typeof hive.fullName === "string"
              ? hive.fullName
              : "";
        const hivePath = path.isAbsolute(hivePathRaw)
          ? hivePathRaw
          : path.join(AGENT_LEE_ROOT, hivePathRaw);
        if (fs.existsSync(hivePath)) {
          const raw = stripBom(fs.readFileSync(hivePath, "utf8"));
          const parsed = tolerantJsonParse(raw);
          if (parsed && parsed.modelRoles && parsed.modelRoles[roleName]) return parsed.modelRoles[roleName];
        }
      } catch (err) {
        // ignore and fallthrough
      }
    }
  }
  return null;
}

export default {
  loadAllLeewayIndex,
  findModelRole
};
